apisix/plugins/proxy-chain.lua

```diff
@@ -0,0 +1,182 @@
+-- Proxy Chain Plugin for APISIX
+-- Author: Vahid Aghazadeh v.opsource@gmail.com
+-- Description: This plugin chains multiple upstream service requests, merging their responses into a single payload.
+-- It supports passing a custom token header (e.g., Authorization) between services for authentication purposes.
+-- License: Apache License 2.0
+--
+-- Licensed to the Apache Software Foundation (ASF) under one or more
+-- contributor license agreements.  See the NOTICE file distributed with
+-- this work for additional information regarding copyright ownership.
+-- The ASF licenses this file to You under the Apache License, Version 2.0
+-- (the "License"); you may not use this file except in compliance with
+-- the License.  You may obtain a copy of the License at
+--
+--     http://www.apache.org/licenses/LICENSE-2.0
+--
+-- Unless required by applicable law or agreed to in writing, software
+-- distributed under the License is distributed on an "AS IS" BASIS,
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+--
+
+local core = require("apisix.core")        -- Core APISIX utilities
+local http = require("resty.http")         -- HTTP client for making service calls
+local cjson = require("cjson")             -- JSON encoding/decoding library
+
+local plugin_name = "proxy-chain"
+
+-- Schema definition for plugin configuration
+local schema = {
+    type = "object",
+    properties = {
+        services = {
+            type = "array",
+            items = {
+                type = "object",
+                properties = {
+                    uri = { type = "string", minLength = 1 },          -- URI of the service to call
+                    method = { type = "string", enum = {"GET", "POST", "PUT", "DELETE"}, default = "POST" }  -- HTTP method
+                },
+                required = {"uri"}  -- URI is mandatory
+            },
+            minItems = 1  -- At least one service must be specified
+        },
+        token_header = { type = "string" }  -- Optional header name for passing a token
+    },
+    required = {"services"}  -- Services array is mandatory
+}
+
+-- Plugin metadata
+local _M = {
+    version = 0.1,                    -- Plugin version
+    priority = 1000,                  -- Execution priority (higher runs earlier)
+    name = plugin_name,               -- Plugin name
+    schema = schema,                  -- Configuration schema
+    description = "A plugin to chain multiple service requests and merge their responses."
+}
+
+-- Validate the plugin configuration against the schema
+function _M.check_schema(conf)
+    return core.schema.check(schema, conf)
+end
+
+-- Access phase: Chain service calls and merge responses
+function _M.access(conf, ctx)
+    -- Read the incoming request body
+    ngx.req.read_body()
+    local original_body = ngx.req.get_body_data()
```

Comment:
Please note that when the request body exceeds the Nginx buffer size, it will be written to a disk file, at which point this function will return nil. You need to also check ngx.req.get_body_file() to ensure that all request bodies are read correctly. You can refer to how other plugins (such as body-transformer) handle this.